### PR TITLE
_check_array_lengths properly handles corner cases with None

### PR DESCRIPTION
**Pull request by ahundt**
small part of #6928 and #7046

### CODE REVIEW

File: keras/engine/training.py [hunk 1]
```diff
@@ -211,29 +211,29 @@ def _check_array_lengths(inputs, targets, weights):
     # Raises
         ValueError: in case of incorrectly formatted data.
     """
-    x_lengths = [x.shape[0] for x in inputs]
-    y_lengths = [y.shape[0] for y in targets]
-    w_lengths = [w.shape[0] for w in weights]
-    set_x = set(x_lengths)
+    def set_of_lengths(x):
+        # return a set with the variation between of different shapes, with None => 0
```
Comment (fchollet) [inline keras/engine/training.py]: Fix comment

File: keras/engine/training.py [hunk 2]
```diff
@@ -211,29 +211,29 @@ def _check_array_lengths(inputs, targets, weights):
...
+        return set([0]) if x is None else set([0 if y is None else y.shape[0] for y in x])
+    set_x = set_of_lengths(inputs)
```
Comment (fchollet) [inline keras/engine/training.py]: Add blank line above

File: keras/engine/training.py [hunk 3]
```diff
@@ -211,29 +211,29 @@ def _check_array_lengths(inputs, targets, weights):
     # Raises
         ValueError: in case of incorrectly formatted data.
     """
-    x_lengths = [x.shape[0] for x in inputs]
-    y_lengths = [y.shape[0] for y in targets]
-    w_lengths = [w.shape[0] for w in weights]
-    set_x = set(x_lengths)
+    def set_of_lengths(x):
+        # return a set with the variation between of different shapes, with None => 0
+        return set([0]) if x is None else set([0 if y is None else y.shape[0] for y in x])
```
Comment (fchollet) [inline keras/engine/training.py]: Break into `if/else` blocks, for readability

File: keras/engine/training.py [hunk 4]
```diff
@@ -211,29 +211,34 @@ def _check_array_lengths(inputs, targets, weights):
...
     """
-    x_lengths = [x.shape[0] for x in inputs]
-    y_lengths = [y.shape[0] for y in targets]
-    w_lengths = [w.shape[0] for w in weights]
-    set_x = set(x_lengths)
+    def set_of_lengths(x):
+        # return a set with the variation between
+        # different shapes, with None => 0
+        if x is None:
+            return set([0])
```
Comment (fchollet) [inline keras/engine/training.py]: This is just `{0}`

Comment (ahundt) [inline keras/engine/training.py]: cool! I didn't know that

**Comment by fchollet** [comment]
LGTM

**Pull request merged**